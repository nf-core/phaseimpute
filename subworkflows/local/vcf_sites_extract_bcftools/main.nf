include { BCFTOOLS_VIEW                 } from '../../../modules/nf-core/bcftools/view'
include { BCFTOOLS_INDEX                } from '../../../modules/nf-core/bcftools/index'
include { TABIX_BGZIP                   } from '../../../modules/nf-core/tabix/bgzip'
include { TABIX_TABIX                   } from '../../../modules/nf-core/tabix/tabix'
include { BCFTOOLS_QUERY                } from '../../../modules/nf-core/bcftools/query'
include { GAWK                          } from '../../../modules/nf-core/gawk'
include { BCFTOOLS_CONVERT              } from '../../../modules/nf-core/bcftools/convert'


workflow VCF_SITES_EXTRACT_BCFTOOLS {
    take:
    ch_vcf          // channel: [ [id, chr], vcf, index ]
    ch_fasta        // channel: [ [genome], fasta, fai ]

    main:

    ch_versions = Channel.empty()
    ch_fasta = ch_fasta.map { meta, fasta, fai -> [meta, fasta] }

    // Convert VCF to Hap and Legend files
    BCFTOOLS_CONVERT(ch_vcf, ch_fasta, [])
    ch_versions = ch_versions.mix(BCFTOOLS_CONVERT.out.versions)

    // Output hap and legend files
    ch_hap_legend = BCFTOOLS_CONVERT.out.hap.join(BCFTOOLS_CONVERT.out.legend)

    // Extract sites positions
    BCFTOOLS_VIEW(ch_vcf, [], [], [])
    ch_versions = ch_versions.mix(BCFTOOLS_VIEW.out.versions.first())

    // Index extracted sites
    BCFTOOLS_INDEX(BCFTOOLS_VIEW.out.vcf)
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX.out.versions.first())

    // Join extracted sites and index
    ch_panel_sites = BCFTOOLS_VIEW.out.vcf.combine(BCFTOOLS_INDEX.out.csi, by:0)

    // Convert to TSV with structure for Glimpse
    BCFTOOLS_QUERY(ch_panel_sites, [], [], [])
    ch_versions = ch_versions.mix(BCFTOOLS_QUERY.out.versions.first())

    // Compress TSV
    TABIX_BGZIP(BCFTOOLS_QUERY.out.output)
    ch_versions = ch_versions.mix(TABIX_BGZIP.out.versions.first())

    // Generate default posfile (sites vcf, sites index and sites txt)
    ch_posfile = ch_panel_sites
            .join(TABIX_BGZIP.out.output)

    // Convert TSV to Stitch format ","" to "\t"
    GAWK(BCFTOOLS_QUERY.out.output, [])
    ch_versions = ch_versions.mix(GAWK.out.versions)

    // Generate glimpse posfile
    ch_glimpse_posfile = ch_posfile.map{ metaPC, sites, s_index, tsv -> [metaPC, sites, tsv]}

    emit:
    hap_legend             = ch_hap_legend       // channel: [ [id, chr], '.hap', '.legend' ]
    panel_tsv_stitch       = GAWK.out.output     // channel: [ [id, chr], txt ]
    panel_sites            = ch_panel_sites      // channel: [ [id, chr], vcf, csi ]
    posfile                = ch_posfile          // channel: [ [id, chr], vcf, csi, tsv.gz ]
    glimpse_posfile        = ch_glimpse_posfile  // channel: [ [id, chr], vcf, tsv.gz ]
    versions               = ch_versions         // channel: [ versions.yml ]
}
