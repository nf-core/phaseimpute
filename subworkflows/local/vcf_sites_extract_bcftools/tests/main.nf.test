nextflow_workflow {

    name "Test Subworkflow VCF_SITES_EXTRACT_BCFTOOLS"
    script "../main.nf"

    config "./nextflow.config"

    workflow "VCF_SITES_EXTRACT_BCFTOOLS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_sites_extract_bcftools"
    tag "vcf_sites_extract_bcftools"

    tag "bcftools"
    tag "bcftools/convert"
    tag "bcftools/view"


    test("Extract sites") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [id: "1000GP", chr: "chr22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                    ],
                    [
                        [id: "1000GP", chr: "chr21"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz.csi", checkIfExist:true),
                    ]
                )
                input[1] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.posfile.collect{
                    path(it[1]).vcf.summary
                    }).match("Extract sites content")
                }
            )
        }
    }
}
