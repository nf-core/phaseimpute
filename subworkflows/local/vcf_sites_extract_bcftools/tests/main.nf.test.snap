{
    "Normalize content": {
        "content": [
            [
                "VcfFile [chromosomes=[chr21], sampleCount=0, variantCount=836, phased=true, phasedAutodetect=true]",
                "VcfFile [chromosomes=[chr22], sampleCount=0, variantCount=903, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T17:18:11.757404002"
    },
    "Extract sites content": {
        "content": [
            [
                "VcfFile [chromosomes=[chr21], sampleCount=0, variantCount=836, phased=true, phasedAutodetect=true]",
                "VcfFile [chromosomes=[chr22], sampleCount=0, variantCount=903, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T17:19:04.507605451"
    },
    "Extract sites": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "1000GP_chr21_glimpse1_sites.vcf.gz:md5,9eff24bfee06f8fe274922ad06ccaf59",
                        "1000GP_chr21_glimpse1_sites.vcf.gz.tbi:md5,92f4dca4b3ed85b53e88c2ce6caa1ba4",
                        "1000GP_chr21.hap.gz:md5,64d7fdfc96ce79848cf9c3c220e6774f",
                        "1000GP_chr21.legend.gz:md5,3f7edc4c153521a0871c9d51f7da2372"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr22"
                        },
                        "1000GP_chr22_glimpse1_sites.vcf.gz:md5,2bd01ad6efeb5e33fc9cc0fb59c5ca4d",
                        "1000GP_chr22_glimpse1_sites.vcf.gz.tbi:md5,89c2c5662e869750dbe0044993b66cb3",
                        "1000GP_chr22.hap.gz:md5,9500b831dcfee86a98577dde9e2cce73",
                        "1000GP_chr22.legend.gz:md5,82f6f97f61e68b4872cc14c0125ff037"
                    ]
                ],
                "1": [
                    "versions.yml:md5,4200660d750728ce3b7ee1f3bce1e18e",
                    "versions.yml:md5,4200660d750728ce3b7ee1f3bce1e18e",
                    "versions.yml:md5,a68010a14b99e45c800704d0ebc37d0d"
                ],
                "posfile": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "1000GP_chr21_glimpse1_sites.vcf.gz:md5,9eff24bfee06f8fe274922ad06ccaf59",
                        "1000GP_chr21_glimpse1_sites.vcf.gz.tbi:md5,92f4dca4b3ed85b53e88c2ce6caa1ba4",
                        "1000GP_chr21.hap.gz:md5,64d7fdfc96ce79848cf9c3c220e6774f",
                        "1000GP_chr21.legend.gz:md5,3f7edc4c153521a0871c9d51f7da2372"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr22"
                        },
                        "1000GP_chr22_glimpse1_sites.vcf.gz:md5,2bd01ad6efeb5e33fc9cc0fb59c5ca4d",
                        "1000GP_chr22_glimpse1_sites.vcf.gz.tbi:md5,89c2c5662e869750dbe0044993b66cb3",
                        "1000GP_chr22.hap.gz:md5,9500b831dcfee86a98577dde9e2cce73",
                        "1000GP_chr22.legend.gz:md5,82f6f97f61e68b4872cc14c0125ff037"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4200660d750728ce3b7ee1f3bce1e18e",
                    "versions.yml:md5,4200660d750728ce3b7ee1f3bce1e18e",
                    "versions.yml:md5,a68010a14b99e45c800704d0ebc37d0d"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T17:18:09.474680524"
    }
}